zebra-chain/src/orchard/keys.rs

```diff
@@ -0,0 +1,1009 @@
+    /// https://zips.z.cash/protocol/protocol.pdf#concreteprfs
+    #[allow(non_snake_case)]
+    fn from(fvk: FullViewingKey) -> Self {
+        let mut M: BitVec<Lsb0, u8> = BitVec::new();
+
+        M.append(&mut BitVec::<Lsb0, u8>::from_slice(
+            &<[u8; 32]>::from(fvk.spend_validating_key)[..],
+        ));
+        M.append(&mut BitVec::<Lsb0, u8>::from_slice(
+            &<[u8; 32]>::from(fvk.nullifier_deriving_key)[..],
+        ));
+
+        // Commit^ivk_rivk
+        let commit_x = sinsemilla_short_commit(
+            fvk.ivk_commit_randomness.into(),
+            b"z.cash:Orchard-CommitIvk",
+            &M,
+        );
+
+        Self {
+            // TODO: handle the network better, maybe an enum variant constraint?
+            network: Network::default(),
+            // mod r_P
+            scalar: pallas::Scalar::from_bytes(&commit_x.into()).unwrap(),
+        }
+    }
+}
+
+impl FromStr for IncomingViewingKey {
+    type Err = SerializationError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match bech32::decode(s) {
+            Ok((hrp, bytes, Variant::Bech32)) => {
+                let decoded = Vec::<u8>::from_base32(&bytes).unwrap();
+
+                let mut scalar_bytes = [0u8; 32];
+                scalar_bytes[..].copy_from_slice(&decoded[0..32]);
+
+                Ok(IncomingViewingKey {
+                    network: match hrp.as_str() {
+                        ivk_hrp::MAINNET => Network::Mainnet,
+                        _ => Network::Testnet,
+                    },
+                    scalar: pallas::Scalar::from_bytes(&scalar_bytes).unwrap(),
+                })
+            }
+            _ => Err(SerializationError::Parse("bech32 decoding error")),
+        }
+    }
+}
+
+impl PartialEq<[u8; 32]> for IncomingViewingKey {
+    fn eq(&self, other: &[u8; 32]) -> bool {
+        self.scalar.to_bytes() == *other
+    }
+}
+
+/// Magic human-readable strings used to identify what networks Orchard full
+/// viewing keys are associated with when encoded/decoded with bech32.
+///
+/// https://zips.z.cash/protocol/nu5.pdf#orchardfullviewingkeyencoding
+mod fvk_hrp {
+    pub const MAINNET: &str = "zviewo";
+    pub const TESTNET: &str = "zviewtestorchard";
+}
+
+/// Full Viewing Keys
+///
+/// Allows recognizing both incoming and outgoing notes without having
+/// spend authority.
+///
+/// For incoming viewing keys on the production network, the
+/// Human-Readable Part is “zviewo”. For incoming viewing keys on the
+/// test network, the Human-Readable Part is “zviewtestorchard”.
+///
+/// https://zips.z.cash/protocol/protocol.pdf#orchardfullviewingkeyencoding
+#[derive(Copy, Clone, Eq, PartialEq)]
+pub struct FullViewingKey {
+    network: Network,
+    spend_validating_key: SpendValidatingKey,
+    nullifier_deriving_key: NullifierDerivingKey,
+    ivk_commit_randomness: IvkCommitRandomness,
+}
+
+// TODO: impl a From that accepts a Network?
```

Comment:
`from_string(... , network: Network)` ?

Reply:
@teor2345 what do you think of 
```
impl From<(String/Key, Network)> for $type
```

vs

```
impl $type {
    pub fn from_string/key(string:String, network: Network) {
    }
}
```

for these sorts of things?

Reply:
`from_string` is more discoverable, particularly if you're using rustdoc or code completion.